### PR TITLE
Precompute portal node graphs

### PR DESCRIPTION
**Pull request by jere8184**
WIP

### CODE REVIEW

**Comment by jere8184** [comment]
@heinezen, So i've made a graph with weights equal to the distance between two neighboring portals and used this to avoid having to calculate distance in portal_a_star, but the performance is actually turning out to be worse. This is probably due to the weights being incorrect between portals which connect the same two sectors.

i.e. portals 17 and 18, both connect sector A  to sector B. the value returned from distance cost will change distance_cost depending on which sector_center you use.

**Comment by heinezen** [comment]
@jere8184 I'll try to take a look tomorrow. In theory, performance shouldn't actually suffer, but let's see.

File: libopenage/pathfinding/pathfinder.cpp [hunk 1]
```diff
@@ -268,18 +272,38 @@ const Pathfinder::portal_star_t Pathfinder::portal_a_star(const PathRequest &req
 		// get the exits of the current node
 		auto exits = current_node->get_exits(visited_portals, current_node->entry_sector);
 
+		auto graph_node = map[current_node->portal->get_id()];
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.cpp]: You are still not using the precomputed node graph fully. Notice that `PortalNode::get_exits` still creates instances of `PortalNode` which is exactly what we want to avoid. If we want to make it faster, you should try to find all instances where a `PortalNode` is currently created and remove or replace it with code that accesses the precomputed node graph instead.

File: libopenage/pathfinding/grid.cpp [hunk 2]
```diff
@@ -64,6 +64,39 @@ const std::vector<std::shared_ptr<Sector>> &Grid::get_sectors() const {
...
+
+void Grid::init_portal_graph()
+{
+	//map
+	for(auto& sector: this->get_sectors())
+	{
+		for(auto& portal : sector->get_portals())
+		{
+			//node_t portal_node = std::make_shared<PortalNode>(portal, sector->get_id(), nullptr);
+			std::map<portal_id_t, int> links = this->portal_graph[portal->get_id()];
```
Comment (heinezen) [inline libopenage/pathfinding/grid.cpp]: Is there a reason why you used  `map` and not `unordered_map` here? `map` has linear lookup time, while `unordered_map`'s lookup time is constant, so the latter would be faster.

File: libopenage/pathfinding/grid.h [hunk 3]
```diff
@@ -95,6 +98,12 @@ class Grid {
 	 */
 	void init_portals();
 
+	using portalgraph_t = std::unordered_map<portal_id_t, std::map<portal_id_t, int>>;
```
Comment (heinezen) [inline libopenage/pathfinding/grid.h]: The portal node graph should contain the `PortalNode` instances, not just a mapping between IDs. The creation of `PortalNode` instances is what actually takes the most amount of time.

File: libopenage/pathfinding/grid.cpp [hunk 4]
```diff
@@ -64,6 +64,39 @@ const std::vector<std::shared_ptr<Sector>> &Grid::get_sectors() const {
...
+			for(auto& exit : portal->get_exits(sector->get_id()))
+			{
+				//node_t exit_node = std::make_shared<PortalNode>(exit, portal->get_exit_sector(portal_node->entry_sector), nullptr);
+				int cost = Pathfinder::distance_cost(
+								portal->get_exit_center(sector->get_id()),
+								exit->get_entry_center(portal->get_exit_sector(sector->get_id())));
```
Comment (heinezen) [inline libopenage/pathfinding/grid.cpp]: Precomputing the distance cost is a pretty good idea! However, the distance cost should probably be stored in the `PortalNode` class.

**Review by heinezen — CHANGES_REQUESTED** [review]
Hey, I think your solution already goes into the rigt direction, but you might have misunderstood what should be stored in the portal node graph. Maybe we have not explained it well enough in the issue.

- The portal node graph should store the `PortalNode` instances. The creation of these instances is pretty expensive, so we only want to create them once. The type can probably be that of `visited_portals` (i.e. `nodemap_t`).
- Whenever the A\* algorithm wants a `PortalNode`, it should look it up in the node graph and not create it on-the-fly.
- Possible exits and their distance cost can and should be cached in the `PortalNode` object.

File: libopenage/pathfinding/grid.cpp (hunk 4 above)
Comment (jere8184) [reply libopenage/pathfinding/grid.cpp]: @heinezen hey thanks for the all the advice, I will try to address the comments.

File: libopenage/pathfinding/grid.cpp (hunk 2 above)
Comment (jere8184) [reply libopenage/pathfinding/grid.cpp]: Hey, my thinking was that an ordered map would provide the closest neighbor first and this might provide some performance benefit. This would have to be tested ofcourse.

Comment (heinezen) [reply libopenage/pathfinding/grid.cpp]: Generally, I would recommend to use the most viable default type first (e.g. unordered map) and later measure alternatives for speed. Otherwise you might end up with a lot of premature optimization which could make the code slower even.

A vector of these items, sorted by closest neighbor, might also work a little better than a map for speed, since it's memory is contiguos. But let's try that later.

**Comment by jere8184** [comment]
I've observed a 17% increase in speed, with my current changes.

**Comment by heinezen** [comment]
I'm going  to take another look.

File: libopenage/pathfinding/grid.cpp [hunk 5]
```diff
@@ -98,4 +100,32 @@ void Grid::init_portals() {
...
+void Grid::init_portal_nodes() 
+{
+	//create portal_nodes
+	for(auto& sector : this->sectors)
+	{
+		for(auto& portal : sector->get_portals())
+		{
+			if(portal_nodes.find(portal->get_id()) == portal_nodes.end())
+			{
+				portal_nodes[portal->get_id()] = std::make_shared<PortalNode>(portal, NULL, nullptr);
```
Comment (heinezen) [inline libopenage/pathfinding/grid.cpp]: NULL/nullptr should not be passed to the portal node constructor if we only set them later. I think you can add a constructor that just takes the `portal` argument.

Comment (heinezen) [inline libopenage/pathfinding/grid.cpp]: You should use `this` before class members to differentiate them from local members.

```suggestion
			if(this->portal_nodes.find(portal->get_id()) == this->portal_nodes.end())
			{
				this->portal_nodes[portal->get_id()] = std::make_shared<PortalNode>(portal, NULL, nullptr);
```

File: libopenage/pathfinding/grid.cpp [hunk 6]
```diff
@@ -98,4 +100,32 @@ void Grid::init_portals() {
...
+			}
+		}
+	}
+
+	//init portal_node exits
+	for(auto& sector : this->sectors)
+	{
+		for(auto& portal : sector->get_portals())
+		{
+			portal_nodes[portal->get_id()]->init_exits(portal_nodes, sector->get_id());
```
Comment (heinezen) [inline libopenage/pathfinding/grid.cpp]: Same here

```suggestion
			this->portal_nodes[portal->get_id()]->init_exits(this->portal_nodes, sector->get_id());
```

File: libopenage/pathfinding/grid.h [hunk 7]
```diff
@@ -94,6 +96,10 @@ class Grid {
 	 * This should be called after all sectors' cost fields have been initialized.
 	 */
 	void init_portals();
+	
+	const nodemap_t& get_portal_map();
+
+	void init_portal_nodes();
```
Comment (heinezen) [inline libopenage/pathfinding/grid.h]: These methods should have docstrings.

File: libopenage/pathfinding/grid.h [hunk 8]
```diff
@@ -115,6 +121,8 @@ class Grid {
 	 * Sectors of the grid.
 	 */
 	std::vector<std::shared_ptr<Sector>> sectors;
+
+	nodemap_t portal_nodes;
```
Comment (heinezen) [inline libopenage/pathfinding/grid.h:135]: This member should also have a docstring.

File: libopenage/pathfinding/pathfinder.cpp [hunk 9]
```diff
@@ -246,7 +250,7 @@ const Pathfinder::portal_star_t Pathfinder::portal_a_star(const PathRequest &req
 	while (not node_candidates.empty()) {
 		auto current_node = node_candidates.pop();
 
-		current_node->was_best = true;
+		was_best.insert(current_node->portal->get_id());
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.cpp]: Why did you make `was_best` a set instead of storing it in the node? Just curious.

File: libopenage/pathfinding/pathfinder.cpp [hunk 10]
```diff
@@ -266,20 +270,18 @@ const Pathfinder::portal_star_t Pathfinder::portal_a_star(const PathRequest &req
...
 				continue;
 			}
 
-			// Get distance cost (from current node to exit node)
-			auto distance_cost = Pathfinder::distance_cost(
-				current_node->portal->get_exit_center(current_node->entry_sector),
-				exit->portal->get_entry_center(exit->entry_sector));
 
-			bool not_visited = !visited_portals.contains(exit->portal->get_id());
+			bool not_visited = visited_portals.find(exit->portal->get_id()) == visited_portals.end();
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.cpp]: The change is probably not faster if you don't use the result of `find`, so I would change it back for readability.

File: libopenage/pathfinding/pathfinder.h [hunk 11]
```diff
@@ -226,6 +234,13 @@ class PortalNode : public std::enable_shared_from_this<PortalNode> {
 	 * Priority queue node that contains this path node.
 	 */
 	heap_t::element_t heap_node;
+
+
+	using neighbours_t = std::unordered_map<portal_id_t, int>; //I wanted to use weak_ptr as key here but couldnt get it to compile
+
+	std::unordered_map<sector_id_t, neighbours_t> exits;
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.h]: It might be better/faster to not nest the unordered maps, i.e. to make two separate unordered maps.

- One mapping to get the exits (sector ID -> portal ID) :  `std::unordered_map<sector_id_t, portal_id_t>`
- Another mapping to get the distance cost to each exit (portal ID -> cost): `std::unordered_map<portal_id_t, int>`

This will result in another lookup in the second map, but it will probably still be faster since the maps will be closer to each other in memory.

Comment (heinezen) [inline libopenage/pathfinding/pathfinder.h]: You should also document the members.

File: libopenage/pathfinding/pathfinder.h [hunk 12]
```diff
@@ -226,6 +234,13 @@ class PortalNode : public std::enable_shared_from_this<PortalNode> {
 	 * Priority queue node that contains this path node.
 	 */
 	heap_t::element_t heap_node;
+
+
+	using neighbours_t = std::unordered_map<portal_id_t, int>; //I wanted to use weak_ptr as key here but couldnt get it to compile
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.h]: `weak_ptr` is not hashable because the contained object may expire. Also, using `portal_id_t` is probably the right call here anyway :)

**Review by heinezen — CHANGES_REQUESTED** [review]
A few minor points, but it already looks pretty refined.

File: libopenage/pathfinding/pathfinder.cpp (hunk 9 above)
Comment (jere8184) [reply libopenage/pathfinding/pathfinder.cpp]: Because the nodes would have to be reset after each call to portal_a_star, which would be inefficient. As far as I can tell anyway.

Comment (heinezen) [reply libopenage/pathfinding/pathfinder.cpp]: Good point.

File: libopenage/pathfinding/pathfinder.cpp (hunk 10 above)
Comment (jere8184) [reply libopenage/pathfinding/pathfinder.cpp]: i've changed visited_portals to be of type unordered_set, `find` seems to be the best available method. Maybe I could use `insert `and query the result.

File: libopenage/pathfinding/pathfinder.h (hunk 11 above)
Comment (jere8184) [reply libopenage/pathfinding/pathfinder.h:241]: I think there is an issue with the second map.

say we have portals 1 and 2 that both connect sectors A and B, the distance between portals 1 and 2 may vary depending on which sector we enter from. 

That was part of the thinking behind this change.

File: libopenage/pathfinding/pathfinder.cpp (hunk 10 above)
Comment (heinezen) [reply libopenage/pathfinding/pathfinder.cpp]: unordered_set supports `contains()` https://en.cppreference.com/w/cpp/container/unordered_set/contains

File: libopenage/pathfinding/pathfinder.h (hunk 11 above)
Comment (jere8184) [reply libopenage/pathfinding/pathfinder.h]: How would this handle portals, which connect the same two sectors? Maybe I'm missing something.

File: libopenage/pathfinding/pathfinder.cpp [hunk 13]
```diff
@@ -266,20 +270,18 @@ const Pathfinder::portal_star_t Pathfinder::portal_a_star(const PathRequest &req
...
 				continue;
 			}
 
-			// Get distance cost (from current node to exit node)
-			auto distance_cost = Pathfinder::distance_cost(
-				current_node->portal->get_exit_center(current_node->entry_sector),
-				exit->portal->get_entry_center(exit->entry_sector));
 
-			bool not_visited = !visited_portals.contains(exit->portal->get_id());
+			bool not_visited = visited_portals.contains(exit->portal->get_id());
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.cpp]: you are now missing the `!` for testing the negative condition.

File: libopenage/pathfinding/pathfinder.h [hunk 14]
```diff
@@ -181,6 +185,12 @@ class PortalNode : public std::enable_shared_from_this<PortalNode> {
 	 * Get all exits of a node.
 	 */
 	std::vector<node_t> get_exits(const nodemap_t &nodes, sector_id_t entry_sector);
```
Comment (heinezen) [inline libopenage/pathfinding/pathfinder.h]: you can remove this function now that it's no longer used

File: libopenage/pathfinding/pathfinder.h (hunk 11 above)
Comment (heinezen) [reply libopenage/pathfinding/pathfinder.h]: I thought about this again and I think this is something we can just improve on later.

**Review by heinezen — CHANGES_REQUESTED** [review]
Final change requests I hope :D

**Comment by jere8184** [comment]
I'm now seeing a performance decrease. I've messed around with valgrind and from a quick glance I think the map look ups are causing this. :/.

edit: I've got a plan.

**Comment by jere8184** [comment]
@heinezen feel free to take a look, pretty decent performance increase observed :)

**Comment by heinezen** [comment]
Nice, I think this should be ready now :) We can always improve the code later.

I have just a few requests before merging:

- You should format you code with `clamg-format` one last time (and I still can really recommend to turn autoformat on in whatever IDE you are using)
- Since https://github.com/SFTtech/openage/pull/1656 got merged, the code from the target is now in the openage main repo, so it would be good to reopen the PR there (and target `master`)
